### PR TITLE
[113] Show current user's rating on starability

### PR DESCRIPTION
**Pull request by czorek**

### CODE REVIEW

**Comment by czorek** [comment]
[Enable modifying rate's value (re-rating)](https://trello.com/c/P6dG5woq/73-enable-modifying-rate-s-value-re-rating)

**Comment by czorek** [comment]
[[113] Show current user's rate on the rating stars if there is one](https://trello.com/c/qnKBMRBJ/74-113-show-current-user-s-rate-on-the-rating-stars-if-there-is-one)

File: app/services/rate_checker.rb [hunk 1]
```diff
@@ -1,7 +1,20 @@
...
+    @submission = Submission.find(submission_id)
+  end
 
+  def user_has_already_rated?
     submission.rates.any? { |rate| rate.user_id == user_id }
   end
+
+  def current_user_rate_value
+    rate = Rate.find_by(submission_id: submission.id, user_id: user_id)
+    return 0 unless rate
```
Comment (przemyslawhaptas) [inline app/services/rate_checker.rb]: `rate ? rate.value : 0`

File: app/services/rate_creator.rb [hunk 2]
```diff
@@ -2,7 +2,8 @@ class RateCreator
   def self.build(value, submission_id, user_id)
     submission = Submission.find(submission_id)
     user = User.find(user_id)
-    new(value, submission, user, RateChecker.new)
+    rate_checker = RateChecker.new(submission.id, user.id)
```
Comment (przemyslawhaptas) [inline app/services/rate_creator.rb]: I'm not sure if build it here or pass a ready RateChecker object to the `build` method. @lesniakania ?

File: app/services/rate_creator.rb [hunk 3]
```diff
@@ -13,17 +14,17 @@ def initialize(value, submission, user, rate_checker)
...
   def call
-    rate = Rate.new({ value: @value, submission: @submission, user: @user })
-
-    if @rate_checker.user_has_already_rated?(@submission.id, @user.id)
-        errors = ['You are not allowed to rate the same submission twice']
-        success = false
-        result = Result.new(rate, success, errors)
+    if rate_checker.user_has_already_rated?
+      rate = Rate.find_by(user_id: user.id, submission_id: submission.id)
+      success = rate.update_attribute(:value, value)
```
Comment (przemyslawhaptas) [inline app/services/rate_creator.rb]: What about `rate.value = value` here and move `success = rate.save` outside `if/else`?

File: app/services/rate_checker.rb [hunk 4]
```diff
@@ -1,7 +1,20 @@
 class RateChecker
-  def user_has_already_rated?(submission_id, user_id)
-    submission = Submission.find(submission_id)
+  def initialize(submission_id, user_id)
```
Comment (przemyslawhaptas) [inline app/services/rate_checker.rb]: After the conversation about DI I'd like to inject ready objects in the constructor and parameters to `build` methods, what do you think?

File: app/views/submissions/show.html.erb [hunk 5]
```diff
@@ -107,7 +107,7 @@
 
     <aside class="aside-widget aside-widget-absolute">
       <%= form_tag(submission_rate_path(submission), class: "submission-rating-form") do %>
-        <fieldset class="starability-fade fieldset-inline">
+        <fieldset class="starability-fade fieldset-inline" id="user-rating" data-rate="<%= current_user_rate_value %>">
```
Comment (przemyslawhaptas) [inline app/views/submissions/show.html.erb]: It's a nice custom to scope an attribute that's only a javascript handler with `js-` (here: `id="js-user-rating"`)

File: spec/services/rate_creator_spec.rb [hunk 6]
```diff
@@ -1,13 +1,27 @@
...
-  it 'creates a new rate' do
-    result = subject.call
-    expect(Rate.where(value: 1, submission_id: submission.id, user_id: user.id).first).not_to be_nil
-    expect(result.success).to be true
+  let(:submission) { FactoryGirl.create(:submission) }
+  let(:user) { FactoryGirl.create(:user) }
+  let(:value) { 1 }
+  subject { described_class.build(value, submission.id, user.id) }
+
+  context 'when no rates' do
```
Comment (przemyslawhaptas) [inline spec/services/rate_creator_spec.rb]: I'd write something more explicit here like "when the user hasn't rated the submission yet", similarly in the second context

File: spec/services/rate_creator_spec.rb [hunk 7]
```diff
@@ -1,13 +1,27 @@
...
+    it 'creates a new rate' do
+      result = subject.call
+      expect(Rate.where(value: value, submission_id: submission.id, user_id: user.id).first).not_to be_nil
```
Comment (przemyslawhaptas) [inline spec/services/rate_creator_spec.rb]:
```Ruby
expect(Rate.find_by(value: value, submission_id: submission.id, user_id: user.id)).not_to be_nil
expect { result }.to change(Rate, :count).by(1)
```

File: app/services/rate_checker.rb (hunk 4 above)
Comment (czorek) [inline app/services/rate_checker.rb]: I think that's alright

File: app/services/rate_creator.rb [hunk 8]
```diff
@@ -13,17 +14,17 @@ def initialize(value, submission, user, rate_checker)
   end
 
   def call
-    rate = Rate.new({ value: @value, submission: @submission, user: @user })
-
-    if @rate_checker.user_has_already_rated?(@submission.id, @user.id)
-        errors = ['You are not allowed to rate the same submission twice']
-        success = false
-        result = Result.new(rate, success, errors)
+    if rate_checker.user_has_already_rated?
```
Comment (przemyslawhaptas) [inline app/services/rate_creator.rb]: Now that I think about it... Do we need this `user_has_already_rated?` and the conditional at all?
Couldn't we do it in these three lines:
```Ruby
def call
  rate = Rate.find_or_create(user_id: user.id, submission_id: submission.id)
  success = rate.update(value: value)
  Result.new(rate, success)
end
```

Comment (przemyslawhaptas) [inline app/services/rate_creator.rb]: And if there is not other use of the method we could delete it from the `RateChecker`

**Pull request merged**